Automattic/wp-calypso
# Signup: Handle user selecting site or domain in `domain-first` flow

This PR:

- Adds the `SiteOrDomain` step, which allows the user to specify if they want a site or just a domain in the `domain-first` signup flow.
- Refactors `goToNextStep` and `goToStep` in `Signup`. Previously, `goToStep` didn't handle animating the scroll position.
- Updates `NavigationLink` to handle skipping steps when going back, which we need here because the user must skip the `themes` and `plans` steps if they decide to purchase a domain instead of creating a site. @scruffian and I discussed a few different ways we could handle this, such as redirecting the user to a different flow (which I implemented yesterday), but the only way to preserve the functionality of the back button that I could come up with was to support skipping steps within the signup framework itself with a `wasSkipped` property. If more flows in the future decide to add this functionality, we might want to create an action for it or even make this configurable in `signup/config`.
- Renames `signupProgressStore` to `signupProgress` as this prop contains the progress data itself, not the store.

**Note:** I would recommend reviewing the commits themselves because [the last one](https://github.com/Automattic/wp-calypso/pull/10599/commits/ffd22090cd4572d617f5aa6055fbccecba0eceef) simply replaces a string throughout `client/signup`, which makes the aggregate diff a little harder to follow.

**Testing**
*Purchasing a domain*
- Visit `/start/domain-first` in a new session (e.g. incognito).
- Submit a domain search and select a domain.
- Select `Just buy a domain`.
- Assert that you are taken to the `user` step.
- Click `BACK`.
- Assert that you are taken back to the `site-or-domain` step.
- Select `Just buy a domain`.
- Enter valid information for a new user and submit.
- Assert that you are taken to checkout with the domain you selected in the second step in your cart.

*Starting a site*
- Visit `/start/domain-first` in a new session (e.g. incognito).
- Submit a domain search and select a domain.
- Select `Start a new site`.
- Select a theme.
- Select a plan.
- Click `BACK`.
- Assert that you are taken back to the `plans` step.
- Select a plan again.
- Enter valid information for a new user and submit.
- Assert that you are taken to checkout with the domain and plan you selected in your cart.

*The main flow*
Some of the changes here are to the `Signup` component itself, so we should test the `main` flow to assert that it still works properly.
- Visit `/start` in a new session (e.g. incognito).
- Assert that you can complete signup through this flow, landing in `checkout` or on your new site depending on whether you selected a domain/plan.
